group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java

```diff
@@ -65,191 +94,233 @@ public String name() {
-                    List<Integer> partitionsToAssign = unassignedPartitionsForTopic
-                        .subList(unassignedPartitionsListStartPointer, unassignedPartitionsListStartPointer + remaining);
-                    unassignedPartitionsListStartPointer += remaining;
-                    newTargetAssignment.computeIfAbsent(memberId, k -> new MemberAssignmentImpl(new HashMap<>()))
-                        .partitions()
-                        .computeIfAbsent(topicId, k -> new HashSet<>())
-                        .addAll(partitionsToAssign);
-                }
+    /**
+     * Sorts the member Ids in the group based on their instance Id if present, otherwise by member Id.
+     * This is done to ensure that the relative ordering of members doesn't change with static members
+     * thus resulting in a sticky assignment.
+     *
+     * @param groupSpec     The group specification containing the member information.
+     * @return a sorted list of member Ids.
+     */
+    private List<String> sortMemberIds(
+        GroupSpec groupSpec
+    ) {
+        List<String> sortedMemberIds = new ArrayList<>(groupSpec.memberIds());
+        Map<String, Optional<String>> instanceIdCache = new HashMap<>();
+
+        for (String memberId : sortedMemberIds) {
+            instanceIdCache.put(memberId, groupSpec.memberSubscription(memberId).instanceId());
+        }
+
+        sortedMemberIds.sort((memberId1, memberId2) -> {
+            Optional<String> instanceId1 = instanceIdCache.get(memberId1);
+            Optional<String> instanceId2 = instanceIdCache.get(memberId2);
+
+            if (instanceId1.isPresent() && instanceId2.isPresent()) {
+                return instanceId1.get().compareTo(instanceId2.get());
+            } else if (instanceId1.isPresent()) {
+                return -1;
+            } else if (instanceId2.isPresent()) {
+                return 1;
+            } else {
+                return memberId1.compareTo(memberId2);
             }
         });
+        return sortedMemberIds;
+    }
 
-        return new GroupAssignment(newTargetAssignment);
+    /**
+     * Assigns a range of partitions to the specified topic based on the provided metadata.
+     *
+     * @param topicMetadata         Metadata containing the topic details, including the number of partitions,
+     *                              the next range to assign, minQuota, and extra partitions.
+     * @param memberAssignment      Map from topic Id to the set of assigned partition Ids.
+     */
+    private void addPartitionsToAssignment(
+        TopicMetadata topicMetadata,
+        Map<Uuid, Set<Integer>> memberAssignment
+    ) {
+        if (topicMetadata.nextRange >= topicMetadata.numPartitions) {
+            memberAssignment.put(topicMetadata.topicId, Collections.emptySet());
+        } else {
+            int start = topicMetadata.nextRange;
+            int end = Math.min(start + topicMetadata.minQuota, topicMetadata.numPartitions);
```

Comment:
I wonder if we should rather adjust the quota before computing the end in order to really ensure that we never go above numPartitions. In practice, it should never append if the assignment was computed correctly until the last member. What do you think?

Reply:
yeah makes sense, made the changes!